src/fsharp/IlxGen.fs

```diff
@@ -2133,8 +2139,27 @@ let DoesGenExprStartWithSequencePoint g sp expr =
 //-------------------------------------------------------------------------
 // Generate expressions
 //-------------------------------------------------------------------------
+let rec GenExpr cenv cgbuf eenv sp (expr: Expr) sequel =
+    cenv.exprRecursionDepth <- cenv.exprRecursionDepth + 1
+
+    if cenv.exprRecursionDepth > 1 then
+        StackGuard.EnsureSufficientExecutionStack cenv.exprRecursionDepth
+        GenExprAux cenv cgbuf eenv sp expr sequel
+    else
+        GenExprWithStackGuard cenv cgbuf eenv sp expr sequel
+
+    cenv.exprRecursionDepth <- cenv.exprRecursionDepth - 1
+
+and GenExprWithStackGuard cenv cgbuf eenv sp expr sequel =
+    assert (cenv.exprRecursionDepth = 1)
+    try
+        GenExprAux cenv cgbuf eenv sp expr sequel
+        assert (cenv.exprRecursionDepth = 1)
+    with
+    | :? System.InsufficientExecutionStackException ->
+        reraise ()
```

Comment:
Either we get rid of the `try .. with` or show a more specific error when we run out of stack space.